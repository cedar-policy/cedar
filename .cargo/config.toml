[target.'cfg(all())']
rustflags = [
    "-Dclippy::unwrap_used",
    "-Dclippy::expect_used",
    "-Dclippy::fallible_impl_from",
    "-Dclippy::unreachable",
    "-Dclippy::indexing_slicing",
    "-Dclippy::panic",
]
