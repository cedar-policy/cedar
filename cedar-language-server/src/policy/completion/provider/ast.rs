/*
 * Copyright Cedar Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::sync::Arc;

use cedar_policy_core::{
    ast::{BinaryOp, EntityType, Expr, ExprVisitor, Literal, Pattern},
    parser::Loc,
};
use smol_str::SmolStr;

use crate::{
    policy::{
        context::{
            AttrContext, AttrContextKind, BinaryOpContext, InKind, IsContext, Op, ReceiverContext,
        },
        CompletionContextKind, DocumentContext,
    },
    utils::is_cursor_in_condition_braces,
};

/// A visitor implementation for gathering completion information from Cedar policy expressions.
///
/// This visitor traverses the Cedar AST to identify the expression context at the cursor position
/// within policy condition blocks(the `when/unless {...}` section of a policy) and determines what kind
/// of completions should be offered to the user.
pub(crate) struct ConditionCompletionVisitor<'a> {
    cx: &'a DocumentContext<'a>,
}

impl<'a> ConditionCompletionVisitor<'a> {
    /// Creates a new completion visitor with the given document context.
    fn new(document_context: &'a DocumentContext<'_>) -> Self {
        Self {
            cx: document_context,
        }
    }

    /// Analyzes the current policy and cursor position to determine the appropriate completion context.
    ///
    /// This is the main entry point for obtaining completion suggestions based on where the
    /// cursor is positioned within a Cedar policy's conditions.
    pub(crate) fn get_completion_context(
        document_context: &DocumentContext<'_>,
    ) -> CompletionContextKind {
        let expr = document_context.policy.non_scope_constraints();
        if !is_cursor_in_condition_braces(
            document_context.cursor_position,
            document_context.policy_text,
        ) {
            return CompletionContextKind::None;
        }

        let mut visitor = ConditionCompletionVisitor::new(document_context);
        let kind = visitor
            .visit_expr(expr)
            .unwrap_or(CompletionContextKind::Unknown);

        let word = document_context
            .get_word_under_cursor()
            // Filters out the typing of entity euid literals as to not provide suggestions for it
            .filter(|w| !w.contains(':'));

        match kind {
            // Must type a character to get suggestions within empty when blocks or after a completion
            CompletionContextKind::Unknown if word.is_some() => {
                CompletionContextKind::ConditionDefault
            }
            kind => kind,
        }
    }

    fn visit_attr(
        &mut self,
        expr: &Arc<Expr>,
        attr: &SmolStr,
        loc: Option<&Loc>,
        context: AttrContextKind,
    ) -> Option<<Self as ExprVisitor>::Output> {
        self.visit_expr(expr).or_else(|| {
            let word = self.cx.get_word_under_cursor();
            // If cursor is within attribute and matches the current attribute name (or empty attribute)
            if self.cx.is_cursor_over_loc(loc) && (attr.is_empty() || word == Some(attr)) {
                return CompletionContextKind::Attr(AttrContext::new(
                    ReceiverContext::new(expr.clone()),
                    context,
                ))
                .into();
            }
            None
        })
    }
}

impl ExprVisitor for ConditionCompletionVisitor<'_> {
    type Output = CompletionContextKind;

    fn visit_literal(
        &mut self,
        lit: &cedar_policy_core::ast::Literal,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        if let Some(loc) = loc {
            // Check if cursor is within the literal
            if self.cx.is_cursor_over_loc(loc) {
                // We're specifically interested in EntityUID literals for action completions
                if let Literal::EntityUID(entity_uid) = lit {
                    let word = self.cx.get_word_under_cursor();
                    if word == Some(entity_uid.eid().escaped().as_str()) {
                        return CompletionContextKind::EntityLiteral {
                            entity_uid: entity_uid.clone(),
                        }
                        .into();
                    }
                }

                if let Literal::String(..) = lit {
                    return CompletionContextKind::None.into();
                }
            }
        } else {
            // Fall back to entity ID location for action completion in policy scope
            if let Literal::EntityUID(entity_uid) = lit {
                return self
                    .visit_literal(&Literal::EntityUID(entity_uid.clone()), entity_uid.loc());
            }
        }
        None
    }

    fn visit_is(
        &mut self,
        expr: &Arc<Expr>,
        entity_type: &EntityType,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        self.visit_expr(expr).or_else(|| {
            if let Some(loc) = loc {
                if self.cx.is_cursor_over_loc(loc) {
                    let word = self.cx.get_word_under_cursor();
                    if word == Some(&entity_type.to_string()) {
                        return CompletionContextKind::Is(IsContext::new(ReceiverContext::new(
                            expr.clone(),
                        )))
                        .into();
                    }
                }
            }
            None
        })
    }

    fn visit_binary_op(
        &mut self,
        op: cedar_policy_core::ast::BinaryOp,
        arg1: &Arc<Expr>,
        arg2: &Arc<Expr>,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        // Try to get completions from either operand first
        self.visit_expr(arg1)
            .or_else(|| self.visit_expr(arg2))
            .or_else(|| {
                if let Some(loc) = loc {
                    // Determine which side needs completion
                    let complete_right_side = !arg1
                        .source_loc()
                        .is_some_and(|loc| self.cx.is_cursor_over_loc(loc));

                    let (other_side_expr, complete_side_expr) = if complete_right_side {
                        (arg1.clone(), arg2.clone())
                    } else {
                        (arg2.clone(), arg1.clone())
                    };

                    if self.cx.is_cursor_over_loc(loc) {
                        match op {
                            BinaryOp::In => {
                                let kind = if complete_side_expr
                                    .source_loc()
                                    .and_then(|loc| loc.snippet())
                                    .is_some_and(|snip| snip.contains('[') && snip.contains(']'))
                                {
                                    InKind::ArrayLiteral
                                } else {
                                    InKind::Entity
                                };
                                return CompletionContextKind::BinaryOp(BinaryOpContext {
                                    op: Op::In(kind),
                                    other_side_expr,
                                    complete_side_expr,
                                })
                                .into();
                            }
                            BinaryOp::Eq => {
                                return CompletionContextKind::BinaryOp(BinaryOpContext {
                                    op: Op::Eq,
                                    other_side_expr,
                                    complete_side_expr,
                                })
                                .into()
                            }
                            _ => (),
                        }
                    }
                }

                None
            })
    }

    fn visit_has_attr(
        &mut self,
        expr: &Arc<Expr>,
        attr: &SmolStr,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        self.visit_attr(expr, attr, loc, AttrContextKind::Has)
    }

    fn visit_get_attr(
        &mut self,
        expr: &Arc<Expr>,
        attr: &SmolStr,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        self.visit_attr(expr, attr, loc, AttrContextKind::Get)
    }

    fn visit_like(
        &mut self,
        expr: &Arc<Expr>,
        pattern: &Pattern,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        self.visit_expr(expr).or_else(|| {
            let word = self.cx.get_word_under_cursor();
            let pat = pattern.to_string();
            if self.cx.is_cursor_over_loc(loc)
                && (word == Some(pat.as_str()) || word.is_none() && pat.is_empty())
            {
                return CompletionContextKind::None.into();
            }
            None
        })
    }
}
