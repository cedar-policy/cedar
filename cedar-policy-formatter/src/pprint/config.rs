/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use super::token::WrappedToken;

/// Configuraton struct that specifies line width and indentation width
#[derive(Debug, Clone)]
pub struct Config {
    pub line_width: usize,
    pub indent_width: isize,
}

#[derive(Debug)]
pub struct Context<'a> {
    pub config: &'a Config,
    pub tokens: Vec<WrappedToken>,
}
