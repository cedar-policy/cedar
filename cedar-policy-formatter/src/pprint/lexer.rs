/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use super::token::{Comment, Token, WrappedToken};
use logos::Logos;

// Attach comments to tokens
fn add_comments(token_stream: &mut [WrappedToken], input: &str) -> Option<()> {
    let mut i = 0;
    if token_stream.is_empty() {
        return None;
    }
    let first_token = token_stream.first_mut()?;
    let first_range = &first_token.span;
    if first_range.start > 0 {
        first_token.add_leading_comment(input.get(..first_range.start)?);
    }

    while i + 1 < token_stream.len() {
        let (curr_tokens, next_tokens) = token_stream.split_at_mut(i + 1);
        let curr_token = curr_tokens.last_mut()?;
        let next_token = next_tokens.first_mut()?;
        let curr_range = &curr_token.span;
        let next_range = &next_token.span;
        if curr_range.end == next_range.start {
            i += 1;
            continue;
        }
        let gap = input.get(curr_range.end..next_range.start)?;
        match gap.split_once('\n') {
            Some((f, r)) => {
                curr_token.add_trailing_comment(f);
                next_token.add_leading_comment(r);
            }
            None => {
                curr_token.add_trailing_comment(gap);
            }
        }
        i += 1;
    }
    Some(())
}

pub fn get_token_stream(input: &str) -> Option<Vec<WrappedToken>> {
    let mut wrapped_tokens: Vec<WrappedToken> = Token::lexer(input)
        .spanned()
        .map(|(t, s)| Some(WrappedToken::new(t.ok()?, Comment::default(), s)))
        .collect::<Option<Vec<WrappedToken>>>()?;
    add_comments(&mut wrapped_tokens, input);
    Some(wrapped_tokens)
}
