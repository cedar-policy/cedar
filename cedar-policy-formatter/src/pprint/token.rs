/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// PANIC SAFETY: there's little we can do about Logos.
#![allow(clippy::indexing_slicing)]
use itertools::Itertools;
use logos::{Logos, Span};
use regex::Regex;
use smol_str::SmolStr;
use std::fmt::{self, Display};

pub fn get_comment(text: &str) -> String {
    // PANIC SAFETY: this regex pattern is valid
    #[allow(clippy::unwrap_used)]
    let mut comment = Regex::new(r"//[^\n\r]*")
        .unwrap()
        .find_iter(text)
        .map(|c| c.as_str().trim_end())
        .join("\n");
    if comment.is_empty() {
        comment
    } else {
        comment.push('\n');
        comment
    }
}

// Represent Cedar comments
#[derive(Clone, Debug, Default, PartialEq)]
pub struct Comment {
    pub leading_comment: String,
    pub trailing_comment: String,
}

impl Comment {
    pub fn add_leading_comment(&mut self, comment: &str) {
        self.leading_comment.push_str(&get_comment(comment));
    }

    pub fn add_trailing_comment(&mut self, comment: &str) {
        self.trailing_comment.push_str(&get_comment(comment));
    }
}

impl Display for Comment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.leading_comment.is_empty() {
            self.trailing_comment.fmt(f)
        } else {
            write!(f, "{}\n{}", self.leading_comment, self.trailing_comment)
        }
    }
}

// Cedar tokens
#[derive(Logos, Clone, Debug, PartialEq)]
pub enum Token {
    #[regex(r"\s*", logos::skip)]
    Whitespace,

    #[regex(r"//[^\n\r]*[\n\r]*", logos::skip)]
    Comment,

    #[token("true")]
    True,

    #[token("false")]
    False,

    #[token("if")]
    If,

    #[token("permit")]
    Permit,

    #[token("forbid")]
    Forbid,

    #[token("when")]
    When,

    #[token("unless")]
    Unless,

    #[token("in")]
    In,

    #[token("has")]
    Has,

    #[token("like")]
    Like,

    #[token("is")]
    Is,

    #[token("then")]
    Then,

    #[token("else")]
    Else,

    #[token("principal")]
    Principal,

    #[token("action")]
    Action,

    #[token("resource")]
    Resource,

    #[token("context")]
    Context,

    #[token("?principal")]
    PrincipalSlot,

    #[token("?resource")]
    ResourceSlot,

    #[regex(r"[_a-zA-Z][_a-zA-Z0-9]*", |lex| SmolStr::new(lex.slice()))]
    Identifier(SmolStr),

    #[regex("[0-9]+", |lex| SmolStr::new(lex.slice()))]
    Number(SmolStr),

    #[regex(r#""(\\.|[^"\\])*""#, |lex| SmolStr::new(lex.slice()))]
    Str(SmolStr),

    #[token("@")]
    At,

    #[token(".")]
    Dot,

    #[token(",")]
    Comma,

    #[token(";")]
    SemiColon,

    #[token(":")]
    Colon,

    #[token("::")]
    DoubleColon,

    #[token("(")]
    LParen,

    #[token(")")]
    RParen,

    #[token("{")]
    LBrace,

    #[token("}")]
    RBrace,

    #[token("[")]
    LBracket,

    #[token("]")]
    RBracket,

    #[token("==")]
    Equal,

    #[token("!=")]
    NotEqual,

    #[token("<")]
    Lt,

    #[token("<=")]
    Le,

    #[token(">")]
    Gt,

    #[token(">=")]
    Ge,

    #[token("||")]
    Or,

    #[token("&&")]
    And,

    #[token("+")]
    Add,

    #[token("-")]
    Dash,

    #[token("*")]
    Mul,

    #[token("/")]
    Div,

    #[token("%")]
    Modulo,

    #[token("!")]
    Neg,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Action => write!(f, "action"),
            Self::Add => write!(f, "+"),
            Self::And => write!(f, "&&"),
            Self::At => write!(f, "@"),
            Self::Colon => write!(f, ":"),
            Self::Comma => write!(f, ","),
            // PANIC SAFETY: comment should be ignored as specified by the lexer regex
            #[allow(clippy::unreachable)]
            Self::Comment => unreachable!("comment should be skipped!"),
            Self::Context => write!(f, "context"),
            Self::Dash => write!(f, "-"),
            Self::Div => write!(f, "/"),
            Self::Dot => write!(f, "."),
            Self::DoubleColon => write!(f, "::"),
            Self::Else => write!(f, "else"),
            Self::Equal => write!(f, "=="),
            Self::False => write!(f, "false"),
            Self::Forbid => write!(f, "forbid"),
            Self::Ge => write!(f, ">="),
            Self::Gt => write!(f, ">"),
            Self::Has => write!(f, "has"),
            Self::Identifier(i) => write!(f, "{}", i),
            Self::If => write!(f, "if"),
            Self::In => write!(f, "in"),
            Self::LBrace => write!(f, "{{"),
            Self::LBracket => write!(f, "["),
            Self::LParen => write!(f, "("),
            Self::Le => write!(f, "<="),
            Self::Like => write!(f, "like"),
            Self::Is => write!(f, "is"),
            Self::Lt => write!(f, "<"),
            Self::Modulo => write!(f, "%"),
            Self::Mul => write!(f, "*"),
            Self::Neg => write!(f, "!"),
            Self::NotEqual => write!(f, "!="),
            Self::Number(n) => write!(f, "{}", n),
            Self::Or => write!(f, "||"),
            Self::Permit => write!(f, "permit"),
            Self::Principal => write!(f, "principal"),
            Self::PrincipalSlot => write!(f, "principal?"),
            Self::RBrace => write!(f, "}}"),
            Self::RBracket => write!(f, "]"),
            Self::RParen => write!(f, ")"),
            Self::Resource => write!(f, "resource"),
            Self::ResourceSlot => write!(f, "resource?"),
            Self::SemiColon => write!(f, ";"),
            Self::Str(s) => write!(f, "{}", s),
            Self::Then => write!(f, "then"),
            Self::True => write!(f, "true"),
            Self::Unless => write!(f, "unless"),
            Self::When => write!(f, "when"),
            // PANIC SAFETY: whitespace should be ignored as specified by the lexer regex
            #[allow(clippy::unreachable)]
            Self::Whitespace => unreachable!("whitespace should be skipped!"),
        }
    }
}

// A wrapper for token span (i.e., (Token, Span))
// We use this wrapper for easier processing of comments
#[derive(Debug, Clone, PartialEq)]
pub struct WrappedToken {
    pub token: Token,
    pub comment: Comment,
    pub span: Span,
}

impl WrappedToken {
    pub fn new(token: Token, comment: Comment, span: Span) -> Self {
        Self {
            token,
            comment,
            span,
        }
    }

    pub fn add_leading_comment(&mut self, comment: &str) {
        self.comment.add_leading_comment(comment);
    }

    pub fn add_trailing_comment(&mut self, comment: &str) {
        self.comment.add_trailing_comment(comment);
    }

    fn clear_leading_comment(&mut self) {
        self.comment.leading_comment.clear();
    }

    fn clear_trailing_comment(&mut self) {
        self.comment.trailing_comment.clear();
    }

    pub fn consume_leading_comment(&mut self) -> String {
        let comment = self.comment.leading_comment.clone();
        self.clear_leading_comment();
        comment
    }

    pub fn consume_comment(&mut self) -> Comment {
        let comment = self.comment.clone();
        self.clear_leading_comment();
        self.clear_trailing_comment();
        comment
    }
}

impl Display for WrappedToken {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{} {}",
            self.comment.leading_comment, self.token, self.comment.trailing_comment
        )
    }
}
