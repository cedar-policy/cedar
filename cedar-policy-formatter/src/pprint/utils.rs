/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use itertools::Itertools;
use pretty::RcDoc;

use super::token::{Comment, WrappedToken};

// Add brackets
pub fn add_brackets<'a>(d: RcDoc<'a>, leftp: RcDoc<'a>, rightp: RcDoc<'a>) -> RcDoc<'a> {
    leftp.append(d.nest(1)).append(rightp)
}

pub fn get_leading_comment_doc_from_str<'a>(leading_comment: &str) -> RcDoc<'a> {
    if leading_comment.is_empty() {
        RcDoc::nil()
    } else {
        let cs: RcDoc<'_> = RcDoc::intersperse(
            leading_comment
                .trim()
                .split('\n')
                .map(|c| RcDoc::text(c.to_owned())),
            RcDoc::hardline(),
        );
        RcDoc::hardline().append(cs).append(RcDoc::hardline())
    }
}

pub fn get_trailing_comment_doc_from_str<'a>(trailing_comment: &str) -> RcDoc<'a> {
    if trailing_comment.is_empty() {
        RcDoc::nil()
    } else {
        RcDoc::space()
            .append(RcDoc::text(trailing_comment.trim().to_owned()))
            .append(RcDoc::hardline())
    }
}

fn get_token_at_start(
    span: miette::SourceSpan,
    tokens: &mut [WrappedToken],
) -> Option<&mut WrappedToken> {
    tokens
        .as_mut()
        .iter_mut()
        .find(|t| t.span.start == span.offset())
}

pub fn get_comment_at_start(
    span: miette::SourceSpan,
    tokens: &mut [WrappedToken],
) -> Option<Comment> {
    Some(get_token_at_start(span, tokens)?.consume_comment())
}

pub fn get_leading_comment_at_start(
    span: miette::SourceSpan,
    tokens: &mut [WrappedToken],
) -> Option<String> {
    Some(get_token_at_start(span, tokens)?.consume_leading_comment())
}

fn get_token_after_end(
    span: miette::SourceSpan,
    tokens: &mut [WrappedToken],
) -> Option<&mut WrappedToken> {
    let end = span.offset() + span.len();
    tokens.iter_mut().find_or_first(|t| t.span.start >= end)
}

fn get_token_at_end(
    span: miette::SourceSpan,
    tokens: &mut [WrappedToken],
) -> Option<&mut WrappedToken> {
    let end = span.offset() + span.len();
    tokens.iter_mut().find(|t| t.span.end == end)
}

pub fn get_comment_at_end(
    span: miette::SourceSpan,
    tokens: &mut [WrappedToken],
) -> Option<Comment> {
    Some(get_token_at_end(span, tokens)?.consume_comment())
}

pub fn get_comment_after_end(
    span: miette::SourceSpan,
    tokens: &mut [WrappedToken],
) -> Option<Comment> {
    Some(get_token_after_end(span, tokens)?.consume_comment())
}

pub fn get_comment_in_range(span: miette::SourceSpan, tokens: &mut [WrappedToken]) -> Vec<Comment> {
    tokens
        .iter_mut()
        .skip_while(|t| t.span.start < span.offset())
        .take_while(|t| t.span.end <= span.offset() + span.len())
        .map(|t| t.consume_comment())
        .collect()
}

// Wrap doc with comment
pub fn add_comment<'a>(d: RcDoc<'a>, comment: Comment, next_doc: RcDoc<'a>) -> RcDoc<'a> {
    let leading_comment = comment.leading_comment;
    let trailing_comment = comment.trailing_comment;
    let leading_comment_doc = get_leading_comment_doc_from_str(&leading_comment);
    let trailing_comment_doc: RcDoc<'_> = if trailing_comment.is_empty() {
        d.append(next_doc)
    } else {
        d.append(RcDoc::space())
            .append(RcDoc::text(trailing_comment.trim().to_owned()))
            .append(RcDoc::hardline())
    };

    leading_comment_doc.append(trailing_comment_doc.clone())
}

pub fn remove_empty_lines(s: &str) -> String {
    s.lines()
        .filter(|ss| !ss.trim().is_empty())
        .map(|s| s.to_owned())
        .collect::<Vec<String>>()
        .join("\n")
}
