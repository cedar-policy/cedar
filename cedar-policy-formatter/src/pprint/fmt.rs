/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use miette::{miette, Result, WrapErr};

use cedar_policy_core::ast::{PolicySet, Template};
use cedar_policy_core::parser::parse_policyset;
use cedar_policy_core::parser::{err::ParseErrors, text_to_cst::parse_policies};

use crate::token::get_comment;

use super::lexer::get_token_stream;
use super::utils::remove_empty_lines;

use super::config::{self, Config};
use super::doc::*;

fn tree_to_pretty<T: Doc>(t: &T, context: &mut config::Context<'_>) -> Result<String> {
    let mut w = Vec::new();
    let config = context.config;
    let doc = t.to_doc(context);
    doc.ok_or(miette!("failed to produce doc"))?
        .render(config.line_width, &mut w)
        .map_err(|err| miette!(format!("failed to render doc: {err}")))?;
    String::from_utf8(w)
        .map_err(|err| miette!(format!("failed to convert rendered doc to string: {err}")))
}

fn soundness_check(ps: &str, ast: &PolicySet) -> Result<()> {
    let formatted_ast = parse_policyset(ps).wrap_err("formatter produces invalid policies")?;
    let (formatted_policies, policies) = (
        formatted_ast.templates().collect::<Vec<&Template>>(),
        ast.templates().collect::<Vec<&Template>>(),
    );

    if formatted_policies.len() != policies.len() {
        return Err(miette!("missing formatted policies"));
    }

    for (f_p, p) in formatted_policies.into_iter().zip(policies.into_iter()) {
        let (f_anno, anno) = (
            f_p.annotations()
                .collect::<std::collections::HashMap<_, _>>(),
            p.annotations().collect::<std::collections::HashMap<_, _>>(),
        );
        if !(f_anno == anno
            && f_p.effect() == p.effect()
            && f_p.principal_constraint() == p.principal_constraint()
            && f_p.action_constraint() == p.action_constraint()
            && f_p.resource_constraint() == p.resource_constraint()
            && f_p
                .non_head_constraints()
                .eq_shape(p.non_head_constraints()))
        {
            return Err(miette!(
                "policies differ:\nformatted: {}\ninput: {}",
                f_p,
                p
            ));
        }
    }
    Ok(())
}

pub fn policies_str_to_pretty(ps: &str, config: &Config) -> Result<String> {
    let cst = parse_policies(ps).wrap_err("cannot parse input policies to CSTs")?;
    let mut errs = ParseErrors::new();
    let ast = cst
        .to_policyset(&mut errs)
        .ok_or(errs)
        .wrap_err("cannot parse input policies to ASTs")?;
    let tokens = get_token_stream(ps).ok_or(miette!("cannot get token stream"))?;
    let end_comment_str = ps
        .get(
            tokens
                .last()
                .ok_or(miette!("token stream is empty"))?
                .span
                .end..,
        )
        .ok_or(miette!("cannot get ending comment string"))?;
    let mut context = config::Context { config, tokens };
    let mut formatted_policies = cst
        .as_inner()
        .ok_or(miette!("fail to get input policy CST"))?
        .0
        .iter()
        .map(|p| Ok(remove_empty_lines(tree_to_pretty(p, &mut context)?.trim())))
        .collect::<Result<Vec<String>>>()?
        .join("\n\n");
    // handle comment at the end of a policyset
    let (trailing_comment, end_comment) = match end_comment_str.split_once('\n') {
        Some((f, r)) => (get_comment(f), get_comment(r)),
        None => (get_comment(end_comment_str), String::new()),
    };
    match (trailing_comment.as_ref(), end_comment.as_ref()) {
        ("", "") => {}
        (_, "") => {
            formatted_policies.push(' ');
            formatted_policies.push_str(&trailing_comment);
        }
        ("", _) => {
            formatted_policies.push('\n');
            formatted_policies.push_str(&end_comment);
        }
        _ => {
            formatted_policies.push(' ');
            formatted_policies.push_str(&trailing_comment);
            formatted_policies.push_str(&end_comment);
        }
    };
    // add soundness check to make sure formatting doesn't alter policy ASTs
    soundness_check(&formatted_policies, &ast)?;
    Ok(formatted_policies)
}

#[cfg(test)]
mod tests {
    use super::*;
    const TEST_CONFIG: &Config = &Config {
        line_width: 40,
        indent_width: 2,
    };

    #[test]
    fn trivial_permit() {
        let policy = r#"permit (principal, action, resource);"#;
        assert_eq!(policies_str_to_pretty(policy, TEST_CONFIG).unwrap(), policy);
    }

    #[test]
    fn trivial_forbid() {
        let policy = r#"forbid (principal, action, resource);"#;
        assert_eq!(policies_str_to_pretty(policy, TEST_CONFIG).unwrap(), policy);
    }

    #[test]
    fn action_in_set() {
        let policy = r#"permit (
        principal in UserGroup::"abc",
        action in [Action::"viewPhoto", Action::"viewComments"],
        resource in Album::"one"
      );"#;
        assert_eq!(
            policies_str_to_pretty(policy, TEST_CONFIG).unwrap(),
            r#"permit (
  principal in UserGroup::"abc",
  action in
    [Action::"viewPhoto",
     Action::"viewComments"],
  resource in Album::"one"
);"#
        );
    }

    #[test]
    fn test_format_files() {
        use std::fs::read_to_string;
        use std::path::Path;

        let config = Config {
            line_width: 80,
            indent_width: 2,
        };
        let dir_path = Path::new(env!("CARGO_MANIFEST_DIR")).join("tests");
        let pairs = vec![
            ("test.cedar", "test_formatted.cedar"),
            ("policies.cedar", "policies_formatted.cedar"),
            ("is_policies.cedar", "is_policies_formatted.cedar"),
        ];
        for (pf, ef) in pairs {
            // editors or cargo run try to append a newline at the end of files
            // we should remove them for equality testing
            assert_eq!(
                policies_str_to_pretty(&read_to_string(dir_path.join(pf)).unwrap(), &config)
                    .unwrap()
                    .trim_end_matches('\n'),
                read_to_string(dir_path.join(ef))
                    .unwrap()
                    .trim_end_matches('\n')
            );
        }
    }
}
