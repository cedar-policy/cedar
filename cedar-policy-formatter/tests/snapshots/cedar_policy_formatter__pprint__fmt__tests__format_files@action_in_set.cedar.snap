---
source: cedar-policy-formatter/src/pprint/fmt.rs
expression: formatted
input_file: cedar-policy-formatter/tests/action_in_set.cedar
---
permit (
  principal in UserGroup::"abc",
  action in [Action::"viewPhoto", Action::"viewComments"],
  resource in Album::"one"
);
